Clone of large repositories into RocksDB fails on Windows

Issue initially reported via the geogig-devel mailing list:

> Hi,
> I am using Geogig 1.0 on Windows 7, Oracle JRE 1.8.0_121. I have a remote repository served via Geoserver/PostGIS, containing approx 1GB of data for Antarctica.
> 
> From the documentation, I think I should be able to clone the remote repository into a directory on the local machine. I tried to do this first by issuing the command:
> 
> geogig.bat clone http://<servername>/geoserver/geogig/repos/master master
> 
> in the intended directory. I get the following output:
> 
> Cloning into 'master'...
> 
> Fetching objects from refs/heads/master
> 
> 10:55:42.835 [main] ERROR o.locationtech.geogig.cli.GeogigCLI - An unhandled error occurred: org.rocksdb.RocksDBException: ▒. See the log for more details.
> java.lang.RuntimeException: org.rocksdb.RocksDBException: ▒
> at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-18.0.jar:na]
> at org.locationtech.geogig.rocksdb.RocksdbObjectStore.putAll(RocksdbObjectStore.java:382) ~[geogig-rocksdb-1.0.jar:1.0]
> at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.putAll(ForwardingObjectDatabase.java:162) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.di.ObjectDatabasePutInterceptor$GraphUpdatingObjectDatabase.putAll(ObjectDatabasePutInterceptor.java:98) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.putAll(ForwardingObjectDatabase.java:162) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.BinaryPackedObjects.ingest(BinaryPackedObjects.java:224) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.HttpRemoteRepo.fetchMoreData(HttpRemoteRepo.java:456) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.HttpRemoteRepo.fetchNewData(HttpRemoteRepo.java:208) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.FetchOp._call(FetchOp.java:257) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.FetchOp._call(FetchOp.java:48) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.repository.AbstractGeoGigOp.call(AbstractGeoGigOp.java:153) ~[geogig-api-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.CloneOp._call(CloneOp.java:163) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.CloneOp._call(CloneOp.java:39) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.repository.AbstractGeoGigOp.call(AbstractGeoGigOp.java:153) ~[geogig-api-1.0.jar:1.0]
> at org.locationtech.geogig.cli.porcelain.Clone.runInternal(Clone.java:164) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.AbstractCommand.run(AbstractCommand.java:68) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.GeogigCLI.executeInternal(GeogigCLI.java:528) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.GeogigCLI.execute(GeogigCLI.java:367) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.app.CLI.main(CLI.java:80) [geogig-cli-app-1.0.jar:1.0]
> Caused by: org.rocksdb.RocksDBException: ▒
> at org.rocksdb.RocksDB.write0(Native Method) ~[rocksdbjni-4.13.4.jar:na]
> at org.rocksdb.RocksDB.write(RocksDB.java:602) ~[rocksdbjni-4.13.4.jar:na]
> at org.locationtech.geogig.rocksdb.RocksdbObjectStore.putAll(RocksdbObjectStore.java:371) ~[geogig-rocksdb-1.0.jar:1.0]
> 
> ... 17 common frames omitted
> 
> An unhandled error occurred: org.rocksdb.RocksDBException: ▒. See the log for more details.
> 
> The error happens consistently after fetching around 300,000 objects (out of about 1,130,000). The logfile gives no extra information over what’s pasted above.
> 
> I have tried:
> 
> geogig.bat init
> 
> in the directory as well, in case it was missing an initialised empty repo, but get the same response. With identical software versions on a Centos 7 Linux Virtual Box VM on the same Windows box, I get:
> 
> geogig clone http://<servername>/geoserver/geogig/repos/master master
> 
> Cloning into 'master'...
> 
> Fetching objects from refs/heads/master
> 
> 1,129,655
> 
> Processed 1,129,697 objects. Inserted: 1,129,697. Existing: 0. Time: 21.46 min. Compressed size: 630,289,302 bytes. Uncompressed size: 1,257,346,509 bytes.
> 
> 100%
> 
> Done.
> 
> The created local repo is as expected. It looks like I might have hit some size or complexity limit in the Windows case? It looks like I am using rocksdb 4.13.4.

I asked David to replace the RocksDB jar (4.13.4) with a slightly older version (4.11.2) to see if the error messages generated were more informative and got this:

> Thanks for your input. I replaced the rocksdb jar with 4.11.2 and repeated the clone – I now get:
> 
> Fetching objects from refs/heads/master
> 
> 17:12:32.171 [main] ERROR o.locationtech.geogig.cli.GeogigCLI - An unhandled error occurred: org.rocksdb.RocksDBException: IO error: Failed to FlushViewOfFile: C:\Users\DARB1.BS-DARB1-L5\Documents\projects\add\data\working\master\.geogig\objects.rocksdb/000015.sst: The process cannot access the file because another process has locked a portion of the file.
> 
> . See the log for more details.
> 
> java.lang.RuntimeException: org.rocksdb.RocksDBException: IO error: Failed to FlushViewOfFile: C:\Users\DARB1.BS-DARB1-L5\Documents\projects\add\data\working\master\.geogig\objects.rocksdb/000015.sst: The process cannot access the file because another process has locked a portion of the file.
> at com.google.common.base.Throwables.propagate(Throwables.java:160) ~[guava-18.0.jar:na]
> at org.locationtech.geogig.rocksdb.RocksdbObjectStore.putAll(RocksdbObjectStore.java:382) ~[geogig-rocksdb-1.0.jar:1.0]
> at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.putAll(ForwardingObjectDatabase.java:162) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.di.ObjectDatabasePutInterceptor$GraphUpdatingObjectDatabase.putAll(ObjectDatabasePutInterceptor.java:98) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.putAll(ForwardingObjectDatabase.java:162) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.BinaryPackedObjects.ingest(BinaryPackedObjects.java:224) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.HttpRemoteRepo.fetchMoreData(HttpRemoteRepo.java:456) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.remote.HttpRemoteRepo.fetchNewData(HttpRemoteRepo.java:208) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.FetchOp._call(FetchOp.java:257) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.FetchOp._call(FetchOp.java:48) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.repository.AbstractGeoGigOp.call(AbstractGeoGigOp.java:153) ~[geogig-api-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.CloneOp._call(CloneOp.java:163) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.porcelain.CloneOp._call(CloneOp.java:39) ~[geogig-core-1.0.jar:1.0]
> at org.locationtech.geogig.repository.AbstractGeoGigOp.call(AbstractGeoGigOp.java:153) ~[geogig-api-1.0.jar:1.0]
> at org.locationtech.geogig.cli.porcelain.Clone.runInternal(Clone.java:164) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.AbstractCommand.run(AbstractCommand.java:68) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.GeogigCLI.executeInternal(GeogigCLI.java:528) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.GeogigCLI.execute(GeogigCLI.java:367) ~[geogig-cli-1.0.jar:1.0]
> at org.locationtech.geogig.cli.app.CLI.main(CLI.java:80) [geogig-cli-app-1.0.jar:1.0]
> Caused by: org.rocksdb.RocksDBException: IO error: Failed to FlushViewOfFile: C:\Users\DARB1.BS-DARB1-L5\Documents\projects\add\data\working\master\.geogig\objects.rocksdb/000015.sst: The process cannot access the file because another process has locked a portion of the file.
> at org.rocksdb.RocksDB.write0(Native Method) ~[rocksdbjni-4.11.2.jar:na]
> at org.rocksdb.RocksDB.write(RocksDB.java:555) ~[rocksdbjni-4.11.2.jar:na]
> at org.locationtech.geogig.rocksdb.RocksdbObjectStore.putAll(RocksdbObjectStore.java:371) ~[geogig-rocksdb-1.0.jar:1.0]
> 
> ... 17 common frames omitted
> 
> An unhandled error occurred: org.rocksdb.RocksDBException: IO error: Failed to FlushViewOfFile: C:\Users\DARB1.BS-DARB1-L5\Documents\projects\add\data\working\master\.geogig\objects.rocksdb/000015.sst: The process cannot access the file because another process has locked a portion of the file.
> 
> . See the log for more details.
> 
> The log contains nothing more than this, apart from three occurrences of:
> 
> 2017-05-03 17:17:07,712 INFO [main] o.l.g.p.CreateDeduplicator [CreateDeduplicator.java:40] No DeduplicationService service found, using default heap based one
> 
> Before the stacktrace. I guess this new error reveals it’s some kind of file locking issue.
> 
> Thanks,
> David.

Hello @emerkle826 , I have a problem that is perhaps similar. Any idea what's going on?

> Exception in thread "main" java.lang.RuntimeException: org.rocksdb.RocksDBException:
        at com.google.common.base.Throwables.propagate(Throwables.java:160)
        at org.locationtech.geogig.rocksdb.RocksConnectionManager.connect(RocksConnectionManager.java:121)
        at org.locationtech.geogig.rocksdb.RocksConnectionManager.connect(RocksConnectionManager.java:34)
        at org.locationtech.geogig.storage.impl.ConnectionManager.acquire(ConnectionManager.java:61)
        at org.locationtech.geogig.rocksdb.RocksdbObjectStore.open(RocksdbObjectStore.java:88)
        at org.locationtech.geogig.rocksdb.RocksdbObjectDatabase.open(RocksdbObjectDatabase.java:70)
        at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.open(ForwardingObjectDatabase.java:40)
        at org.locationtech.geogig.storage.impl.ForwardingObjectDatabase.open(ForwardingObjectDatabase.java:40)
        at org.locationtech.geogig.repository.impl.RepositoryImpl.open(RepositoryImpl.java:116)
        at org.locationtech.geogig.repository.impl.GeoGIG.getRepository(GeoGIG.java:124)
        at org.locationtech.geogig.cli.GeogigCLI.newGeoGIG(GeogigCLI.java:268)
        at org.locationtech.geogig.cli.GeogigCLI.loadRepository(GeogigCLI.java:240)
        at org.locationtech.geogig.cli.GeogigCLI.loadRepository(GeogigCLI.java:235)
        at org.locationtech.geogig.cli.GeogigCLI.getGeogig(GeogigCLI.java:180)
        at org.locationtech.geogig.cli.app.CLI.main(CLI.java:46)
Caused by: org.rocksdb.RocksDBException:
        at org.rocksdb.RocksDB.open(Native Method)
        at org.rocksdb.RocksDB.open(RocksDB.java:286)
        at org.locationtech.geogig.rocksdb.RocksConnectionManager.connect(RocksConnectionManager.java:104)
        ... 13 more

--- a/src/core/src/main/java/org/locationtech/geogig/model/internal/RocksdbHandle.java
+++ b/src/core/src/main/java/org/locationtech/geogig/model/internal/RocksdbHandle.java
@@ -74,12 +74,15 @@ class RocksdbHandle {
     public static RocksdbHandle create(Path targetDir) {
         RocksDB.loadLibrary();
 
+        final String os = System.getProperty("os.name");
+        final boolean isWindows = os.toLowerCase().contains("windows");
+        final boolean safeToUseMMappedFiles = !isWindows;
+
         Options options = new Options();
         options.setCreateIfMissing(true)//
                 .setAdviseRandomOnOpen(true)//
-                .setAllowMmapReads(true)//
-                .setAllowMmapWrites(true)//
-                .setAllowOsBuffer(true)//
+                .setAllowMmapReads(safeToUseMMappedFiles)//
+                .setAllowMmapWrites(safeToUseMMappedFiles)//
                 .setWriteBufferSize(16 * 1024 * 1024)//
                 .setMaxWriteBufferNumber(8)//
                 .setMaxBackgroundCompactions(2)//

--- a/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksConnectionManager.java
+++ b/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksConnectionManager.java
@@ -56,12 +56,15 @@ class RocksConnectionManager extends ConnectionManager<DBConfig, DBHandle> {
 
         RocksDB.loadLibrary();
 
+        final String os = System.getProperty("os.name");
+        final boolean isWindows = os.toLowerCase().contains("windows");
+        final boolean safeToUseMMappedFiles = !isWindows;
+
         org.rocksdb.DBOptions dbOptions = new org.rocksdb.DBOptions();
         dbOptions.setCreateIfMissing(true)//
                 .setAdviseRandomOnOpen(true)//
-                .setAllowMmapReads(true)//
-                .setAllowMmapWrites(true)//
-                .setAllowOsBuffer(true)//
+                .setAllowMmapReads(safeToUseMMappedFiles)//
+                .setAllowMmapWrites(safeToUseMMappedFiles)//
                 .setBytesPerSync(64 * 1024 * 1024);
 
         RocksDB db;

--- a/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksdbObjectStore.java
+++ b/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksdbObjectStore.java
@@ -153,16 +153,15 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
     @Override
     protected boolean putInternal(ObjectId id, byte[] rawData) {
         checkWritable();
-        byte[] key = id.getRawValue();
         boolean exists;
-        exists = exists(bulkReadOptions, key);
-
-        if (!exists) {
-            try (RocksDBReference dbRef = dbhandle.getReference()) {
+        try (RocksDBReference dbRef = dbhandle.getReference()) {
+            byte[] key = id.getRawValue();
+            exists = exists(dbRef, bulkReadOptions, key);
+            if (!exists) {
                 dbRef.db().put(key, rawData);
-            } catch (RocksDBException e) {
-                throw Throwables.propagate(e);
             }
+        } catch (RocksDBException e) {
+            throw Throwables.propagate(e);
         }
         return !exists;
     }
@@ -197,20 +196,20 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
         checkOpen();
         checkNotNull(id, "argument id is null");
 
-        return exists(bulkReadOptions, id.getRawValue());
+        try (RocksDBReference dbRef = dbhandle.getReference()) {
+            return exists(dbRef, bulkReadOptions, id.getRawValue());
+        }
     }
 
     private static final byte[] NO_DATA = new byte[0];
 
-    private boolean exists(ReadOptions readOptions, byte[] key) {
+    private boolean exists(RocksDBReference dbRef, ReadOptions readOptions, byte[] key) {
         int size = RocksDB.NOT_FOUND;
-        try (RocksDBReference dbRef = dbhandle.getReference()) {
-            if (dbRef.db().keyMayExist(key, new StringBuffer())) {
-                try {
-                    size = dbRef.db().get(key, NO_DATA);
-                } catch (RocksDBException e) {
-                    throw Throwables.propagate(e);
-                }
+        if (dbRef.db().keyMayExist(key, new StringBuffer())) {
+            try {
+                size = dbRef.db().get(key, NO_DATA);
+            } catch (RocksDBException e) {
+                throw Throwables.propagate(e);
             }
         }
         return size != RocksDB.NOT_FOUND;
@@ -301,7 +300,7 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
                 while (ids.hasNext()) {
                     ObjectId id = ids.next();
                     id.getRawValue(keybuff);
-                    if (!checkExists || exists(ro, keybuff)) {
+                    if (!checkExists || exists(dbRef, ro, keybuff)) {
                         try {
                             dbRef.db().remove(writeOps, keybuff);
                         } catch (RocksDBException e) {
@@ -354,8 +353,8 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
         Map<ObjectId, Integer> insertedIds = new HashMap<ObjectId, Integer>();
         try (RocksDBReference dbRef = dbhandle.getReference();
                 WriteOptions wo = new WriteOptions()) {
-            wo.setDisableWAL(true);
-            wo.setSync(false);
+             wo.setDisableWAL(true);
+             wo.setSync(false);
             try (ReadOptions ro = new ReadOptions()) {
                 ro.setFillCache(false);
                 ro.setVerifyChecksums(false);
@@ -371,7 +370,7 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
                             object.getId().getRawValue(keybuff);
                             final byte[] value = rawOut.toByteArray();
 
-                            boolean exists = checkExists ? exists(ro, keybuff) : false;
+                            boolean exists = checkExists ? exists(dbRef, ro, keybuff) : false;
                             if (exists) {
                                 listener.found(object.getId(), null);
                             } else {

--- a/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksdbObjectStore.java
+++ b/src/storage/rocksdb/src/main/java/org/locationtech/geogig/rocksdb/RocksdbObjectStore.java
@@ -351,42 +351,44 @@ public class RocksdbObjectStore extends AbstractObjectStore implements ObjectSto
         byte[] keybuff = new byte[ObjectId.NUM_BYTES];
 
         Map<ObjectId, Integer> insertedIds = new HashMap<ObjectId, Integer>();
+
         try (RocksDBReference dbRef = dbhandle.getReference();
-                WriteOptions wo = new WriteOptions()) {
-             wo.setDisableWAL(true);
-             wo.setSync(false);
+                WriteOptions wo = new WriteOptions(); //
+                WriteBatch batch = new WriteBatch()) {
+            wo.setDisableWAL(true);
+            wo.setSync(false);
             try (ReadOptions ro = new ReadOptions()) {
                 ro.setFillCache(false);
                 ro.setVerifyChecksums(false);
                 while (objects.hasNext()) {
                     Iterator<? extends RevObject> partition = Iterators.limit(objects, 10_000);
 
-                    try (WriteBatch batch = new WriteBatch()) {
-                        while (partition.hasNext()) {
-                            RevObject object = partition.next();
-                            rawOut.reset();
-                            writeObject(object, rawOut);
+                    while (partition.hasNext()) {
+                        RevObject object = partition.next();
+                        rawOut.reset();
+                        writeObject(object, rawOut);
 
-                            object.getId().getRawValue(keybuff);
-                            final byte[] value = rawOut.toByteArray();
+                        object.getId().getRawValue(keybuff);
+                        final byte[] value = rawOut.toByteArray();
 
-                            boolean exists = checkExists ? exists(dbRef, ro, keybuff) : false;
-                            if (exists) {
-                                listener.found(object.getId(), null);
-                            } else {
-                                batch.put(keybuff, value);
-                                insertedIds.put(object.getId(), Integer.valueOf(value.length));
-                            }
-
+                        boolean exists = checkExists ? exists(dbRef, ro, keybuff) : false;
+                        if (exists) {
+                            listener.found(object.getId(), null);
+                        } else {
+                            batch.put(keybuff, value);
+                            insertedIds.put(object.getId(), Integer.valueOf(value.length));
                         }
-                        // Stopwatch sw = Stopwatch.createStarted();
-                        dbRef.db().write(wo, batch);
-                        for (Entry<ObjectId, Integer> entry : insertedIds.entrySet()) {
-                            listener.inserted(entry.getKey(), entry.getValue());
-                        }
-                        insertedIds.clear();
-                        // System.err.printf("--- synced writes in %s\n", sw.stop());
+
                     }
+                    // Stopwatch sw = Stopwatch.createStarted();
+                    dbRef.db().write(wo, batch);
+                    batch.clear();
+                    
+                    for (Entry<ObjectId, Integer> entry : insertedIds.entrySet()) {
+                        listener.inserted(entry.getKey(), entry.getValue());
+                    }
+                    insertedIds.clear();
+                    // System.err.printf("--- synced writes in %s\n", sw.stop());
                 }
             }
             wo.sync();

